# Day 1 Reading Journal

Hi everyone.  These are the exercises for the first set of readings from ThinkPython.  While this journal consists of several required exercises, it is meant to encourage active reading more generally.  You should use this journal to take detailed notes, catalog questions, and explore the content from ThinkPython deeply.

For more information on using the Jupyter notebook, check out the documentation [here](http://nbviewer.jupyter.org/github/jupyter/notebook/blob/master/docs/source/examples/Notebook/Examples%20and%20Tutorials%20Index.ipynb).  Additionally, in order to write text in this document, you will be making use of markdown cells. A handy cheatsheet for markdown is available [here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet).

## Chapter 1 - The Way of the Program

Note: the exercise numbers below match up with the reading for ease of cross referencing.

SECTION 1 The Python programming language

High-level languages
1. Python, C, C++, Perl, Java
2. Easy to progeam, less time to write, shorter, easier to read, more likely to be correct, portable (run on different kinds of compyters)

Low-level languages
1. "machine languages" or "assembly languages"
2. Computers can read, shorter processing time

Interpreters: reads high-level program, executes it. Python is an interpreter language
- interactive mode: type python, interpreter displays result
- triple chevron (>>>): indicates that interpreter is ready
- scripts: where code is stored, can be executed in different ways depending on the command window
Compilers: reads program, translates it completely, executes it.

SECTION 2 What is a program?

Program: sequence of instructions, specify how to perform a computation

Input: get ddata from keyboard, a file, other device

Output: display data on screen or send to a file or other device

Math: perform basic operations (addition, mulitplication)

Conditional execution: check for certaincondition, execute appropriate 
code

Repetition: perform some action repeatedly, usually with some variation

SECTION 3.1 Syntax Errors

Issues with the way the program is written. There must be () not ).

If there is a syntax error, python will display and error message and quit

SECTION 3.2 Runtime Errors

Don't appear until after the program has started running. Also called exceptions

SECTION 3.3 Semantic Errors

THe program will run, but it will not be what you wanted it to run

SECTION 3.4 Experimental Debugging

It's important. Do it a lot

SECTION 4 Formal and natural languagaes

Natural language: languages people speak, evolved naturally

Formal languages: languages made for a specific purpose. Math, programming languages.
- strict syntax
- parsing: determining the structuce of the sentence

Ambiguity: formal languages are almost completely unambiguous, regardless of context

Redundancy: formal languages are more concise

Literalness: formal languages are super literal

Poetry: words are used for sounds in natural languages, in formal languages, words are used solely for meaning

Prose: meaning of words is important, but their structuce is more important. More like formal language than poetry, but still often ambiguous

Programs: unambiguougs, literal, can be understoon entirely by the analysis of tokens and structure




SECTION 5 The first program

In [17]:
print 'Hello World'

Hello World


SECTION 6 Debugging

Active reading is good for debugging

Keep it experimental

Prepare for challenges


### Exercise 3

Type `help()` to start the online help utility. Or you can type help('print') to get information about the print statement.  You should type `q` and then hit `enter` in the text box to exit the help utility. 

Note: this exercise is pretty simple (and there's not much to put in the box)!  We just want to make sure that you have tried out this super-handy feature of Python!

In [18]:
help('print')


The ``print`` statement
***********************

   print_stmt ::= "print" ([expression ("," expression)* [","]]
                  | ">>" expression [("," expression)+ [","]])

``print`` evaluates each expression in turn and writes the resulting
object to standard output (see below).  If an object is not a string,
it is first converted to a string using the rules for string
conversions.  The (resulting or original) string is then written.  A
space is written before each object is (converted and) written, unless
the output system believes it is positioned at the beginning of a
line.  This is the case (1) when no characters have yet been written
to standard output, (2) when the last character written to standard
output is a whitespace character except ``' '``, or (3) when the last
write operation on standard output was not a ``print`` statement. (In
some cases it may be functional to write an empty string to standard
output for this reason.)

Note: Objects which act like file objects but

### Exercise 4  

Start the Python interpreter and use it as a calculator. Python’s syntax for math operations is almost the same as standard mathematical notation. For example, the symbols +, - and / denote addition, subtraction and division, as you would expect. The symbol for multiplication is *.

If you run a 10 kilometer race in 43 minutes 30 seconds, what is your average time per mile? What is your average speed in miles per hour? (Hint: there are 1.61 kilometers in a mile). 

In [19]:
10/1.61


6.211180124223602

In [20]:
43.5/6.211180124223602

7.003500000000001

In [21]:
60/7.003500000000001

8.567144998929106

- Average time per mile = 7 minutes
- Average speed in miles per hour = 8.567144998929106

## Chapter 2 - Variables, expression and statements

#### SECTION 1 Values and Types
- string: a string of letter containg within quotation marks
- integer: an integer (don't use commas with large integers, they will be interpreted as a series of integers instead of one large one)
- floating-point: numbers with a decimel

#### SECTION 2 Variables
- variable: a name that refers to a value
- assignment statement: creates new variable and give them values
- state diagram: a way of writing a variable's value

#### SECTION 3 Variable names and keywords
- meaningful variable names document what the variable is used for
- illegal names begin with a number, contain special characters, or are already used by python

#### SECTION 4 Operators and operands
- operators: special symbols, represent computations
- operands: the values the operator is applied to
- +, -, asterisk, /, and double asterisk: perform addition, subtraction, multiplication, division, exponentiation
- if both operands are integers, python performs floor dividion, meaning the decimal truncated
- if you want floating-point, make either operands a floating-point number

#### SECTION 5 Expressions and statements
- expression: combination of values, variables, statements (x + 7)
- statement: unit of code the interpreter can execute. Has no values

#### SECTION 6 Interactive mode and script mode
- interactive: produces output as you go
- script: produces output at the end of the whole thing

#### SECTION 7 Order of operations
- python follows the rules of precedence of math, PEMDAS
- use parentheses to make the order obvious

#### String operations
- no mathematical operations
- concatenation: using + or '*' to connect strings end-to-end (+ adds them and '*' multiplies)

#### Coments
- note, they use '#' to start
- useful for documenting non-obvious portions of the code, like units

#### Debugging
- syntax error messages are not very useful
- runtime errors are usually trying to use a variable before it it defined
- remember PEMDAS

### Exercise 2  

Assume that we execute the following assignment statements:

width = 17
height = 12.0
delimiter = '.'

For each of the following expressions, write the value of the expression and the type (of the value of the expression).

1. `width/2`
2. `width/2.0`
3. `height/3`
4. `1 + 2 * 5`
5. `delimiter * 5`

Please use the following markdown cell to provide your answer.

1. 8
2. 8.5
3. 4.0
4. 11
5. .....

### Exercise 3  

Practice using the Python interpreter as a calculator.  For each of these provide Python code that computes each answer.

    


The volume of a sphere with radius r is 4/3 $\pi r^3$. What is the volume of a sphere with radius 5? Hint: 392.7 is wrong!

In [22]:
radius = 5
pi = 3.1415926535897932
volume = ((4)*pi*(radius**3))/3
print volume

523.598775598


Suppose the cover price of a book is \$24.95, but bookstores get a 40% discount. Shipping costs \$3 for the first copy and 75 cents for each additional copy. What is the total wholesale cost for 60 copies?

In [23]:
orig_book = 24.95
discount = .4
first_book_shipping = 3
other_book_shipping = .75
shipping = first_book_shipping + other_book_shipping*59
sixty_copies = 60*(orig_book*(1-discount)) + shipping
print sixty_copies

945.45


If I leave my house at 6:52 am and run 1 mile at an easy pace (8:15 per mile), then 3 miles at tempo (7:12 per mile) and 1 mile at easy pace again, what time do I get home for breakfast? 

In [24]:
easy_mile = 8.25 #minutes
tempo = 7.2 #minutes
run_length = 3*tempo + 2*easy_mile
time_home = 7.00 + (run_length - 8)*.01
print time_home

7.301


Time home for breakfast = 7.30

## Chapter 3 - Functions

#### SECTION 1 Function calls
- function: a named sequence of statements, performs a computation
- function(argument) = return value

#### SECTION 2 Type conversion functions
- int(): makes integers
- float(): makes floating-point numbers
- str(): makes a string

#### SECTION 3 Math functions
- module: file that contains collection of related functions, must be imported first
- the math module has useful math operations
- to access a function use dot notaion: module.function

#### SECTION 4 Composition
- programs are partially useful for their ability to compose functions and fucntino calls together
- left side must be variable name

#### SECTION 5 Adding new funcitons
- function definition: specifies name of new function and sequence of statements executed when funciton is called
- Example:
        def print_lyrics():
            print "I'm a lumberjack, and I'm okay."
            print "I sleep all night and I work all day."
- function names have the same rules as variable names
- empty () means there are no arguments
- first line of function definition is header, rest is body
- header must end in a colon
- when ending a function definition in interactive mode, there much be an empty line
- you can use defined functions inside of other functions that are being defined

#### SECTION 6 Definition and uses
- functions must be defined, executed, and then called

#### SECTION 7 Flow of execution
- flow of execution: order in which statements are executed, statemetns are executed one at a time from top to bottom
- functions are like detours in execution

#### SECTION 8 Parameters and arguments
- you can pass arguments into functions that you create and that are pre-definied
- name of argument passed through and the name of the parameter do not have to be related

#### SECTION 9 Variables and paramenters are loval
- local: when a variable is created inside of a function it only exists inside of it

#### SECTION 10 Stack diagrams
- useful to keep track of variables and which function they belong to
- frame: represents each function
- order in stack represents which function called which
- traceback: what program file the error occurred in, what line, and what functions were executing at the time

#### SECTION 11 Fruitful functions and void functions
- fruitful functions: yield results
- void funtions: perform action, don't return a value

#### SECTION 12 Why functions?
- breaks things into chunks which makes reading and debugging easier
- eliminates repetitive code
- can be reused

#### SECTION 13 Importing with from
- instead of importing an entire module, you can import just what you want by saying "from module import argument"

#### SECTION 14 Debugging
- tabs and spaces are helpful for debugging and make it look prettier
- some editors are weird about this

### Exercise 3

Python provides a built-in function called len that returns the length of a string, so the value of len('allen') is 5.
Write a function named right_justify that takes a string named s as a parameter and prints the string with enough leading spaces so that the last letter of the string is in column 70 of the display.

```
>>> right_justify('allen')
                                                                 allen```

In [9]:
def right_justify(s):
    spacing = ' ' * (70 - len(s))
    print(spacing + s)
    
    
    
right_justify('allen')

                                                                 allen


### Exercise 5

This exercise can be done using only the statements and other features we have learned so far.

(a) Write a function that draws a grid like the following:
```
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
```
**Hint:** to print more than one value on a line, you can print a comma-separated sequence:
print '+', '-'
If the sequence ends with a comma, Python leaves the line unfinished, so the value printed next appears on the same line.
print '+', 
print '-'
The output of these statements is '+ -'.
A print statement all by itself ends the current line and goes to the next line.

In [11]:
def horizontal():
    print '+ ', '- ' * 4, '+ ', '- ' * 4, '+' 
def vertical():
    print '|',' ' * 9, '|', ' ' * 9, '|'

    
horizontal()
vertical()
vertical()
vertical()
vertical()
horizontal()
vertical()
vertical()
vertical()
vertical()
horizontal()

+  - - - -  +  - - - -  +
|           |           |
|           |           |
|           |           |
|           |           |
+  - - - -  +  - - - -  +
|           |           |
|           |           |
|           |           |
|           |           |
+  - - - -  +  - - - -  +


(b) Write a function that draws a similar grid with four rows and four columns.

In [24]:
def horizontal():
    print '+ ', '- ' * 4, '+ ', '- ' * 4, '+ ', '- ' * 4, '+ ', '- ' * 4, '+' 
def vertical():
    print '|',' ' * 9, '|', ' ' * 9, '|',' ' * 9, '|', ' ' * 9, '|'
def full_vertical():
    vertical()
    vertical()
    vertical()
    vertical()
    
horizontal()
full_vertical()
horizontal()
full_vertical()
horizontal()
full_vertical()
horizontal()
full_vertical()
horizontal()

+  - - - -  +  - - - -  +  - - - -  +  - - - -  +
|           |           |           |           |
|           |           |           |           |
|           |           |           |           |
|           |           |           |           |
+  - - - -  +  - - - -  +  - - - -  +  - - - -  +
|           |           |           |           |
|           |           |           |           |
|           |           |           |           |
|           |           |           |           |
+  - - - -  +  - - - -  +  - - - -  +  - - - -  +
|           |           |           |           |
|           |           |           |           |
|           |           |           |           |
|           |           |           |           |
+  - - - -  +  - - - -  +  - - - -  +  - - - -  +
|           |           |           |           |
|           |           |           |           |
|           |           |           |           |
|           |           |           |           |


In [ ]:
def grid(x, y):
    top_segment = ('+ ' + '- ' * 4) + '\n' #creates the top line of the code
    inner_segment = '|' + ' ' * 9 + '\n' #creates the inner line
    line = top_segment * x + '+' + (inner_segment * x + '|') * 4
    print line * y




******** are the empty spaces the cells???? that's what I went with

## Chapter 5.1 - 5.7 Conditionals

#### Modulus operator
- work on integer and yields the remainder when the 1st operand is divided buy the second
- represented by %
- same syntax as other operators
- finds the remainder when two numbers are divided (if it equals 0, the two numbers are divisible

#### Boolean expressions
- expression that is either true or false
- ==, it is a 'type' like int or float
- relational operators:, != not equal, the others are as expected, > or < comes before =

#### Logial operators
- and: only if both are true
- or: if either are true
- not: negates the boolean expression (is only true if the expression is false)
- any nonzero number is interpreted as true

#### Conditional execution
- conditional statements: checks conditions and changes the behavior of a program accordingly
- if statement: what is after the 'if' is the condition
- if you don't have anything for the body, use pass

#### Alternative execution
- if/else statement
- else statement is a 'branch'

#### Chained conditionals
- if/elif/else statements
- there must be at most one else statement at the very end
- there can be as many elif statements as you want

#### Nested conditionals
- must fufill both if statements


### Exercise 3  

Fermat’s Last Theorem says that there are no positive integers a, b, and c such that $a^n + b^n = c^n$ for any values of n greater than 2.

(a) Write a function named `check_fermat` that takes four parameters—a, b, c and n—and that checks to see if Fermat’s theorem holds. If n is greater than 2 and it turns out to be true that
$a^n + b^n = c^n$ the program should print, "Holy smokes, Fermat was wrong!" Otherwise the program should print, "No, that doesn’t work."

In [30]:
def check_fermat(a, b, c, n):
    if n>2 and a^n + b^n == c^n:
        print 'Holy smokes, Fermal was wrong'
    else:
        print "No, that doesn't work"

(b) Write a function that prompts the user to input values for a, b, c and n, converts them to integers, and uses check_fermat to check whether they violate Fermat’s theorem.

In [35]:
a = input('Input a value for a = ')
b = input('Input a value for b = ')
c = input('Input a value for c = ')
n = input('Input a value for n = ')

check_fermat(a, b, c, n)

Input a value for a = 3
Input a value for b = 4
Input a value for c = 5
Input a value for n = 5
No, that doesn't work


### Exercise 4  

If you are given three sticks, you may or may not be able to arrange them in a triangle. For example, if one of the sticks is 12 inches long and the other two are one inch long, it is clear that you will not be able to get the short sticks to meet in the middle. For any three lengths, there is a simple test to see if it is possible to form a triangle:
> If any of the three lengths is greater than the sum of the other two, then you cannot form a triangle. Otherwise, you can. (If the sum of two lengths equals the third, they form what is called a “degenerate” triangle.)

(a) Write a function named `is_triangle` that takes three integers as arguments, and that prints either "Yes" or "No," depending on whether you can or cannot form a triangle from sticks with the given lengths.

In [38]:
def is_triangle(a, b, c):
    if a + b > c:
        print 'Yes'
    elif a + c > b:
        print 'Yes'
    elif b + c > a:
        print 'Yes'
    else:
        print 'No'

(b) Write a function that prompts the user to input three stick lengths, converts them to integers, and uses is_triangle to check whether sticks with the given lengths can form a triangle.

In [39]:
a = input('Input a value for Side 1 = ')
b = input('Input a value for Side 2 = ')
c = input('Input a value for Side 3 = ')

is_triangle(a, b, c)

Input a value for Side 1 = 3
Input a value for Side 2 = 4
Input a value for Side 3 = 5
Yes


## Notes for the Instructors

Please use the space below to write comments to help us plan the next class session.  For instance, if you want to see us go over an example of a particular type of problem, you can indicate that here.

Please remember that the best way to get quick feedback from the instructors as well as your peers, is to use Piazza.  However, consider using this channel if it doesn't make sense to post your note to Piazza.